File: typescript/src/core/piece.ts
export class Piece {
  static totalFill = 0;

  readonly index: number;
  readonly name: string;
  currRotation = 0;
  row = -1;
  column = -1;
  totalThisFill = 0;

  private readonly firstSquarePos: number[];
  private readonly rowsSet: number[][];
  private readonly columnsSet: number[][];

  constructor(
    index: number,
    layout: number[][],
    availRotations: number,
    symmetric: number,
    name: string,
  ) {
    this.index = index;
    this.name = name;

    const totalOrientations = availRotations * symmetric;
    const layouts: number[][][] = new Array(totalOrientations);
    this.firstSquarePos = new Array(totalOrientations);
    layouts[0] = layout;

    this.rowsSet = new Array(totalOrientations);
    this.columnsSet = new Array(totalOrientations);

    this.firstSquarePos[0] = 0;
    while (
      this.firstSquarePos[0] < layout[0].length &&
      layout[0][this.firstSquarePos[0]] === 0
    ) {
      this.firstSquarePos[0]++;
    }

    let maxColumns = -1;
    for (const row of layout) {
      for (const value of row) {
        if (value === 1) {
          Piece.totalFill++;
          this.totalThisFill++;
        }
      }
      if (row.length > maxColumns) {
        maxColumns = row.length;
      }
    }

    this.printPart(0, layout);
    if (availRotations > 1) {
      layouts[1] = this.realRotate(layouts[0], maxColumns, layout.length, 1);
      this.printPart(1, layouts[1]);
      if (availRotations > 2) {
        layouts[2] = this.realRotate(layouts[1], layout.length, maxColumns, 2);
        this.printPart(2, layouts[2]);
        if (availRotations > 3) {
          layouts[3] = this.realRotate(layouts[2], maxColumns, layout.length, 3);
          this.printPart(3, layouts[3]);
        }
      }
    }

    if (symmetric === 2) {
      for (let i = 0; i < availRotations; i++) {
        layouts[i + availRotations] = this.copySymmetric(layouts[i]);
        this.printPart(i + availRotations, layouts[i + availRotations]);
        this.firstSquarePos[i + availRotations] = 0;
        while (
          this.firstSquarePos[i + availRotations] < layouts[i + availRotations][0].length &&
          layouts[i + availRotations][0][this.firstSquarePos[i + availRotations]] === 0
        ) {
          this.firstSquarePos[i + availRotations]++;
        }
      }
    }

    for (let rot = 0; rot < totalOrientations; rot++) {
      this.rowsSet[rot] = new Array(this.totalThisFill);
      this.columnsSet[rot] = new Array(this.totalThisFill);
      let setSoFar = 0;
      for (let i = 0; i < layouts[rot].length; i++) {
        for (let j = 0; j < layouts[rot][i].length; j++) {
          if (layouts[rot][i][j] === 1) {
            this.rowsSet[rot][setSoFar] = i;
            this.columnsSet[rot][setSoFar] = j;
            setSoFar++;
          }
        }
      }
    }
  }

  getAvailRotations(): number {
    return this.firstSquarePos.length;
  }

  getFirstSquarePos(): number {
    return this.firstSquarePos[this.currRotation];
  }

  getRowSet(i: number): number {
    return this.rowsSet[this.currRotation][i];
  }

  getColumnSet(i: number): number {
    return this.columnsSet[this.currRotation][i];
  }

  getRow(): number {
    return this.row;
  }

  getColumn(): number {
    return this.column;
  }

  rotate(): void {
    this.currRotation++;
    if (this.currRotation === this.firstSquarePos.length) {
      this.currRotation = 0;
    }
  }

  setPosition(row: number, column: number): void {
    this.row = row;
    this.column = column;
  }

  private copySymmetric(original: number[][]): number[][] {
    const rows = original.length;
    const result: number[][] = new Array(rows);
    for (let i = 0; i < rows; i++) {
      result[i] = new Array(original[rows - i - 1].length);
      for (let j = 0; j < result[i].length; j++) {
        result[i][j] = original[rows - i - 1][j];
      }
    }
    return result;
  }

  private realRotate(original: number[][], rows: number, columns: number, index: number): number[][] {
    const result: number[][] = new Array(rows);
    for (let i = 0; i < rows; i++) {
      result[i] = new Array(columns).fill(0);
    }

    for (let i = 0; i < columns; i++) {
      for (let j = 0; j < rows; j++) {
        try {
          result[rows - j - 1][i] = original[i][j];
        } catch (_) {
          // out of bounds
        }
      }
    }

    this.firstSquarePos[index] = 0;
    while (
      this.firstSquarePos[index] < result[0].length &&
      result[0][this.firstSquarePos[index]] === 0
    ) {
      this.firstSquarePos[index]++;
    }

    return result;
  }

  private printPart(l: number, layout: number[][]): void {
    console.log(`    layout=${l}`);
    for (const row of layout) {
      console.log(row.join(''));
    }
  }

  toString(): string {
    return `id ${this.name} rotation ${90 * this.currRotation} used ${this.row !== -1}`;
  }
}

File: python/src/core/piece.py
from __future__ import annotations


class Piece:
    total_fill: int = 0

    def __init__(
        self,
        index: int,
        layout: list[list[int]],
        avail_rotations: int,
        symmetric: int,
        name: str,
    ) -> None:
        self.index = index
        self.name = name
        self.curr_rotation = 0
        self.row = -1
        self.column = -1
        self.total_this_fill = 0

        total_orientations = avail_rotations * symmetric
        layouts: list[list[list[int]]] = [None] * total_orientations  # type: ignore[list-item]
        self.first_square_pos: list[int] = [0] * total_orientations
        layouts[0] = layout

        self.rows_set: list[list[int]] = [None] * total_orientations  # type: ignore[list-item]
        self.columns_set: list[list[int]] = [None] * total_orientations  # type: ignore[list-item]

        self.first_square_pos[0] = 0
        while (
            self.first_square_pos[0] < len(layout[0])
            and layout[0][self.first_square_pos[0]] == 0
        ):
            self.first_square_pos[0] += 1

        max_columns = -1
        for row in layout:
            for value in row:
                if value == 1:
                    Piece.total_fill += 1
                    self.total_this_fill += 1
            if len(row) > max_columns:
                max_columns = len(row)

        self._print_part(0, layout)
        if avail_rotations > 1:
            layouts[1] = self._real_rotate(layouts[0], max_columns, len(layout), 1)
            self._print_part(1, layouts[1])
            if avail_rotations > 2:
                layouts[2] = self._real_rotate(layouts[1], len(layout), max_columns, 2)
                self._print_part(2, layouts[2])
                if avail_rotations > 3:
                    layouts[3] = self._real_rotate(layouts[2], max_columns, len(layout), 3)
                    self._print_part(3, layouts[3])

        if symmetric == 2:
            for i in range(avail_rotations):
                layouts[i + avail_rotations] = self._copy_symmetric(layouts[i])
                self._print_part(i + avail_rotations, layouts[i + avail_rotations])
                self.first_square_pos[i + avail_rotations] = 0
                while (
                    self.first_square_pos[i + avail_rotations]
                    < len(layouts[i + avail_rotations][0])
                    and layouts[i + avail_rotations][0][
                        self.first_square_pos[i + avail_rotations]
                    ]
                    == 0
                ):
                    self.first_square_pos[i + avail_rotations] += 1

        for rot in range(total_orientations):
            self.rows_set[rot] = [0] * self.total_this_fill
            self.columns_set[rot] = [0] * self.total_this_fill
            set_so_far = 0
            for i in range(len(layouts[rot])):
                for j in range(len(layouts[rot][i])):
                    if layouts[rot][i][j] == 1:
                        self.rows_set[rot][set_so_far] = i
                        self.columns_set[rot][set_so_far] = j
                        set_so_far += 1

    def get_avail_rotations(self) -> int:
        return len(self.first_square_pos)

    def get_first_square_pos(self) -> int:
        return self.first_square_pos[self.curr_rotation]

    def get_row_set(self, i: int) -> int:
        return self.rows_set[self.curr_rotation][i]

    def get_column_set(self, i: int) -> int:
        return self.columns_set[self.curr_rotation][i]

    def get_row(self) -> int:
        return self.row

    def get_column(self) -> int:
        return self.column

    def rotate(self) -> None:
        self.curr_rotation += 1
        if self.curr_rotation == len(self.first_square_pos):
            self.curr_rotation = 0

    def set_position(self, row: int, column: int) -> None:
        self.row = row
        self.column = column

    def _copy_symmetric(self, original: list[list[int]]) -> list[list[int]]:
        rows = len(original)
        result: list[list[int]] = [None] * rows  # type: ignore[list-item]
        for i in range(rows):
            result[i] = list(original[rows - i - 1])
        return result

    def _real_rotate(
        self,
        original: list[list[int]],
        rows: int,
        columns: int,
        index: int,
    ) -> list[list[int]]:
        result: list[list[int]] = [[0] * columns for _ in range(rows)]

        for i in range(columns):
            for j in range(rows):
                try:
                    result[rows - j - 1][i] = original[i][j]
                except IndexError:
                    pass

        self.first_square_pos[index] = 0
        while (
            self.first_square_pos[index] < len(result[0])
            and result[0][self.first_square_pos[index]] == 0
        ):
            self.first_square_pos[index] += 1

        return result

    def _print_part(self, l: int, layout: list[list[int]]) -> None:
        print(f"    layout={l}")
        for row in layout:
            print("".join(str(x) for x in row))

    def __str__(self) -> str:
        return f"id {self.name} rotation {90 * self.curr_rotation} used {self.row != -1}"
